core/unit_test/TestCudaCacheConfig.hpp

```diff
@@ -0,0 +1,114 @@
+// notice, this list of conditions and the following disclaimer in the
+// documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY NTESS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL NTESS OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+#include <gtest/gtest.h>
+
+#include <Kokkos_Core.hpp>
+#include <iostream>
+
+// The following unit test is only valid for the CUDA backend.
+#if defined(KOKKOS_ENABLE_CUDA)
+namespace Test {
+TEST(TEST_CATEGORY, cuda_cache_config) {
+  using CachePreference = Kokkos::Impl::CachePreference;
+  CachePreference cache_config;
+  cudaFuncAttributes attributes;
+  cudaDeviceProp prop;
+  cudaGetDeviceProperties(&prop, 0);
+  Kokkos::TeamPolicy<TEST_EXECSPACE> policy_t;
+  dim3 grid{14, 1, 1};
+  dim3 block{64, 10, 1};
+
+  int shmem = 0;
+
+  {
+    attributes.numRegs         = 32;
+    attributes.sharedSizeBytes = 16;
+#if defined(KOKKOS_ARCH_VOLTA70)
+    shmem = 1000;
+#elif defined(KOKKOS_ARCH_AMPERE80)
+    shmem = 10000;
+#else
+    printf(
+        "The unit test is only defined for Volta70 and Ampere80 "
+        "architectures.\n");
+#endif
+
+    modify_launch_configuration_if_desired_occupancy_is_specified(
+        grid, policy_t, prop, attributes, block, shmem, cache_config);
+
+    ASSERT_EQ(cache_config, CachePreference::PreferL1);
+  }
+  {
+    attributes.numRegs         = 64;
+    attributes.sharedSizeBytes = 32;
+#if defined(KOKKOS_ARCH_VOLTA70)
+    shmem = 45000;
+#elif defined(KOKKOS_ARCH_AMPERE80)
+    shmem = 55000;
+#else
+    printf(
+        "The unit test is only defined for Volta70 and Ampere80 "
+        "architectures.\n");
+#endif
+
+    modify_launch_configuration_if_desired_occupancy_is_specified(
+        grid, policy_t, prop, attributes, block, shmem, cache_config);
+
+    ASSERT_EQ(cache_config, CachePreference::PreferEqual);
+  }
+  {
+    attributes.numRegs         = 32;
+    attributes.sharedSizeBytes = 16;
+    shmem                      = 63000;
```

Comment:
I'm surprised this is enough for A100, since the shared memory per SM is configurable up to 164KB... if you have it handy, what does `sharedMemPerMultiprocessor` return for A100?